s3readerat.go

```diff
@@ -16,51 +18,88 @@ import (
-	bucket string
-	key    string
-	size   int64
+	Debug   bool
+	ctx     context.Context
+	client  *s3.Client
+	options *s3.Options
+	bucket  string
+	key     string
+	size    int64
 }
 
-var _ io.ReaderAt = (*S3ReaderAt)(nil)
+type Options struct {
+	// Debug indicates whether to enable debug logging.
+	Debug bool
 
-// New creates a new S3ReaderAt.
-func New(client *s3.Client, bucket string, key string) (ra *S3ReaderAt, err error) {
-	if client == nil {
-		return nil, errors.New("S3 client is required")
-	}
+	// Context is the context.Context to use.
+	Context context.Context
 
-	ra = &S3ReaderAt{
-		ctx:    context.Background(),
-		client: client,
-		bucket: bucket,
-		key:    key,
-	}
+	// Client is the s3.Client to use when running in single-region mode. You can instead pass s3.Options to run in
+	// multi-region mode.
+	Client *s3.Client
 
-	ra.size = -1
-	return ra, nil
+	// Options are the s3.Options to use when running in multi-region mode. In this mode, S3ReaderAt constructs the
+	// s3.Client for you in the appropriate region(s). You can instead pass s3.Client to run in single-region mode.
+	Options *s3.Options
+
+	// Bucket is the AWS S3 bucket to use.
+	Bucket string
+
+	// Key is the key to use within the AWS S3 bucket. It should not start with a leading slash.
+	Key string
+
+	// Size is the size in bytes to use, if known in advance. This is an optimization that avoids calling "HeadObject".
+	Size *int64
+}
+
+var _ io.ReaderAt = (*S3ReaderAt)(nil)
+
+// New creates a new S3ReaderAt.
+func New(client *s3.Client, bucket string, key string) (*S3ReaderAt, error) {
+	return NewWithOptions(Options{
+		Context: context.Background(),
+		Client:  client,
+		Bucket:  bucket,
+		Key:     key,
+	})
 }
 
 // NewWithSize creates a new S3ReaderAt that skips checking the S3 object's size.
-func NewWithSize(client *s3.Client, bucket string, key string, size int64) (ra *S3ReaderAt, err error) {
-	if client == nil {
-		return nil, errors.New("S3 client is required")
-	}
+func NewWithSize(client *s3.Client, bucket string, key string, size int64) (*S3ReaderAt, error) {
+	return NewWithOptions(Options{
+		Context: context.Background(),
+		Client:  client,
+		Bucket:  bucket,
+		Key:     key,
+		Size:    &size,
+	})
+}
 
-	if size < 0 {
-		return nil, errors.Errorf("Provided size is invalid: %d", size)
+func NewWithOptions(options Options) (*S3ReaderAt, error) {
+	if options.Client == nil && options.Options == nil {
+		return nil, errors.New("one of Client or Options is required")
+	} else if options.Client != nil && options.Options != nil {
+		return nil, errors.New("only one of Client or Options can be provided")
+	} else if options.Size != nil && *options.Size < 0 {
+		return nil, errors.Errorf("provided size is invalid: %d", *options.Size)
 	}
-
-	ra = &S3ReaderAt{
-		ctx:    context.Background(),
-		client: client,
-		bucket: bucket,
-		key:    key,
+	ctx := options.Context
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	ra := &S3ReaderAt{
```

Comment:
being picky 😉 , add extra spaces to split contexts
```suggestion

	ctx := options.Context
	if ctx == nil {
		ctx = context.Background()
	}
	
	ra := &S3ReaderAt{
```

Reply:
Good idea